### PR TITLE
Rover: fence support and simple object avoidance

### PR DESCRIPTION
**Pull request by rmackay9**
This PR adds Fence (both circular and polygon) support for Rover and simple object avoidance (stopping before hitting objects) by re-usingthe previously copter specific AC_Avoidance/AC_Avoid library, AC_Fence and AC_Proximity libraries.

- AC_Avoid has been extended to support stopping at the fence instead of always sliding along it.
- AC_Avoid gets a new "adjust_speed" method which is used by Rover to slow the vehicle in Steering and Acro modes before hitting the fence or objects seen by the Proximity library.
- As with Copter, the vehicle will attempt to stop before hitting the circular or polygon fences but if it does breach them, RTL will be triggered.

Some three known issues with this PR:

- the adjust-speed only works properly when ac-avoid is set to Stop (i.e. not slide).  Rovers really can't slide so maybe we should force the stop behaviour regardless of what the user has selected.  As a minimum we should default the behaviour to Stop for Rovers. 
- AC_Avoid relies on the copter specific AC_AttitudeController but only for the static sqrt-controller method.  We could remove this dependency by moving the sqrt-controller to another library.  Any suggestions on where we should put this?  Maybe a new navigation file in AP_Math?
- we should extend the fence breach behaviour to trigger SmartRTL.  Adding SmartRTL to our failsafes is very new though so I think this can be handled in a follow-up PR.

### CODE REVIEW

File: APMrover2/Parameters.h [hunk 1]
```diff
@@ -300,6 +300,7 @@ class ParametersG2 {
     // advanced failsafe library
     AP_AdvancedFailsafe_Rover afs;
 #endif
+
```
Comment (amilcarlucas) [inline APMrover2/Parameters.h]: Whiteline

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 2]
```diff
@@ -409,31 +477,39 @@ void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel,
  * Computes the speed such that the stopping distance
  * of the vehicle will be exactly the input distance.
  */
-float AC_Avoid::get_max_speed(float kP, float accel_cmss, float distance, float dt) const
+float AC_Avoid::get_max_speed(float kP, float accel_cmss, float distance_cm, float dt) const
 {
-    // should use time instead of 0.0f
-    return AC_AttitudeControl::sqrt_controller(distance, kP, accel_cmss, dt);
+    if (is_zero(kP)) {
+        return sqrt(2.0f * distance_cm * accel_cmss);
```
Comment (khancyr) [inline libraries/AC_Avoidance/AC_Avoid.cpp]: I would use safe_sqrt here

File: libraries/AP_Math/vector2.cpp [hunk 3]
```diff
@@ -140,6 +140,104 @@ float Vector2<T>::angle(const Vector2<T> &v2) const
...
 }
 
+// find the intersection between two line segments
+// returns true if they intersect, false if they do not
+// the point of intersection is returned in the intersection argument
+template <typename T>
+bool Vector2<T>::segment_intersection(const Vector2<T>& seg1_start, const Vector2<T>& seg1_end, const Vector2<T>& seg2_start, const Vector2<T>& seg2_end, Vector2<T>& intersection)
+{
+    // implementation borrowed from http://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
+    Vector2<T> r1 = seg1_end - seg1_start;
```
Comment (khancyr) [inline libraries/AP_Math/vector2.cpp]: const

File: libraries/AP_Math/vector2.cpp [hunk 4]
```diff
@@ -140,6 +140,104 @@ float Vector2<T>::angle(const Vector2<T> &v2) const
...
+}
+
+// find the intersection between a line segment and a circle
+// returns true if they intersect and intersection argument is updated with intersection closest to seg_start
+// solution adapted from http://stackoverflow.com/questions/1073336/circle-line-segment-collision-detection-algorithm
+template <typename T>
+bool Vector2<T>::circle_segment_intersection(const Vector2<T>& seg_start, const Vector2<T>& seg_end, const Vector2<T>& circle_center, float radius, Vector2<T>& intersection)
+{
+    // calculate segment start and end as offsets from circle's center
+    Vector2f seg_start_local = seg_start - circle_center;
```
Comment (khancyr) [inline libraries/AP_Math/vector2.cpp]: const

File: libraries/AP_Math/vector2.cpp [hunk 5]
```diff
@@ -140,6 +140,104 @@ float Vector2<T>::angle(const Vector2<T> &v2) const
...
+    //Vector2f seg_end_local = seg_end - circle_center;
+
+    // calculate vector from start to end
+    Vector2f seg_end_minus_start = seg_end - seg_start;
```
Comment (khancyr) [inline libraries/AP_Math/vector2.cpp]: const

**Comment by khancyr** [comment]
@rmackay9 It miss the position check in prearm . This is equivalent to what you call gps_checks in Copter. It just ensure that EKF is ok to allow arming !

File: libraries/AC_Avoidance/AC_Avoid.cpp (hunk 2 above)
Comment (rmackay9) [reply libraries/AC_Avoidance/AC_Avoid.cpp]: yup, will replace with safe_sqrt, txs!

File: libraries/AP_Math/vector2.cpp (hunk 4 above)
Comment (rmackay9) [reply libraries/AP_Math/vector2.cpp]: txs.  will add const.

**Comment by rmackay9** [comment]
I've addressed the items raised by @khancyr (except the position-check item which we can do separately I think) and rebased on master.

**Pull request merged**